sys/fs/squashfs/squashfs_inode.c

```diff
@@ -0,0 +1,266 @@
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ *
+ */
+
+#include <sys/param.h>
+#include <sys/systm.h>
+#include <sys/buf.h>
+#include <sys/conf.h>
+#include <sys/fcntl.h>
+#include <sys/libkern.h>
+#include <sys/limits.h>
+#include <sys/lock.h>
+#include <sys/malloc.h>
+#include <sys/mount.h>
+#include <sys/mutex.h>
+#include <sys/namei.h>
+#include <sys/priv.h>
+#include <sys/proc.h>
+#include <sys/queue.h>
+#include <sys/sbuf.h>
+#include <sys/stat.h>
+#include <sys/uio.h>
+#include <sys/vnode.h>
+
+#include <squashfs.h>
+#include <squashfs_io.h>
+#include <squashfs_mount.h>
+#include <squashfs_block.h>
+#include <squashfs_inode.h>
+#include <squashfs_decompressor.h>
+
+/* Swapendian functions for all types of inodes */
+void		swapendian_base_inode(struct sqsh_base_inode *temp);
+void		swapendian_reg_inode(struct sqsh_reg_inode *temp);
+void		swapendian_lreg_inode(struct sqsh_lreg_inode *temp);
+
+/* init functions for all types of inodes */
+sqsh_err	sqsh_init_reg_inode(struct sqsh_mount *ump, struct sqsh_inode *inode);
+sqsh_err	sqsh_init_lreg_inode(struct sqsh_mount *ump, struct sqsh_inode *inode);
+
+void
+sqsh_metadata_run_inode(struct sqsh_block_run *cur, uint64_t id, off_t base)
+{
+	cur->block = (id >> 16) + base;
+	cur->offset = id & 0xffff;
+}
+
+enum vtype
+sqsh_inode_type(int inode_type)
+{
+	switch (inode_type) {
+		case SQUASHFS_DIR_TYPE:
+		case SQUASHFS_LDIR_TYPE:
+			return (VDIR);
+		case SQUASHFS_REG_TYPE:
+		case SQUASHFS_LREG_TYPE:
+			return (VREG);
+		case SQUASHFS_SYMLINK_TYPE:
+		case SQUASHFS_LSYMLINK_TYPE:
+			return (VLNK);
+		case SQUASHFS_BLKDEV_TYPE:
+		case SQUASHFS_LBLKDEV_TYPE:
+			return (VBLK);
+		case SQUASHFS_CHRDEV_TYPE:
+		case SQUASHFS_LCHRDEV_TYPE:
+			return (VCHR);
+		case SQUASHFS_FIFO_TYPE:
+		case SQUASHFS_LFIFO_TYPE:
+			return (VFIFO);
+		case SQUASHFS_SOCKET_TYPE:
+		case SQUASHFS_LSOCKET_TYPE:
+			return (VSOCK);
+	}
+	return (VBAD);
+}
+
+sqsh_err
+sqsh_verify_inode(struct sqsh_mount *ump, struct sqsh_inode *inode)
+{
+
+	/* check for inode type */
+	if (inode->base.inode_type < 1 || inode->base.inode_type > 14)
+		return (SQFS_ERR);
+
+	/* check for inode_number */
+	if (inode->base.inode_number < 1 || inode->base.inode_number > ump->sb.inodes)
+		return (SQFS_ERR);
+
+	/*
+	 * if inode type is directory then check for parent inode.
+	 * Note that we add +1 because for root inode parent_inode
+	 * is total inodes + 1
+	 */
+	if (inode->base.inode_type == 1) {
+		if (inode->xtra.dir.parent_inode < 1
+			|| inode->xtra.dir.parent_inode > ump->sb.inodes + 1)
+			return (SQFS_ERR);
+	}
+
+	return (SQFS_OK);
+}
+
+uint64_t
+sqsh_root_inode(struct sqsh_mount *ump)
+{
+	return (ump->sb.root_inode);
+}
+
+sqsh_err
+sqsh_get_inode(struct sqsh_mount *ump, struct sqsh_inode *inode,
+	uint64_t id)
+{
+	struct sqsh_block_run cur;
+	sqsh_err err;
+
+	/* initialise all inode fields to 0 */
+	memset(inode, 0, sizeof(*inode));
```

Comment:
is `inode` guaranteed to be non-null?

Reply:
No.
If we encounter an error then we failed to get inode.
I am just initializing its all fields to 0 here for caution that if we still somehow access inode it will not result in segfaults.

Reply:
What happens when `memset()` tries to clear a NULL `inode` pointer?

Reply:
If it is possible for `inode` to be NULL, and the function needs to handle that case, yes, a check and returning an error is good. If, on the other hand, it should not be possible to pass a NULL `inode` (e.g., all functions calling this one would have checked for a NULL `inode` already) adding an ASSERT would be fine.